lib/pbench/agent/tool_meister_client.py

```diff
@@ -0,0 +1,227 @@
+        :to_client_chan: - (optional) a previously constructed
+                           RedisChannelSubscriber object (the caller ensures
+                           it is for the same Redis server client object
+                           given)
+        :logger:         - (optional) a logger to use for reporting any errors
+                           encountered (one will be created if not provided)
+
+        Typically, if you already have a Redis server client object, you would
+        pass that in via "redis_server", otherwise you must pass a host name
+        and port number for the Redis server to use.
+
+        If you already have a RedisChannelSubscriber object, you must provide
+        the Redis server client object that was used to construct it.
+        """
+        assert (
+            redis_server is None and (redis_host is not None and redis_port is not None)
+        ) or (
+            redis_server is not None and (redis_host is None and redis_port is None)
+        ), "You must specify either a redis server object or a redis server host / port pair, but not both"
+
+        self.redis_server = redis_server
+        self.redis_host = redis_host
+        self.redis_port = redis_port
+
+        assert channel_prefix is not None, "You must specify a channel prefix"
+        self.to_client_channel = f"{channel_prefix}-{tm_channel_suffix_to_client}"
+        self.from_client_channel = f"{channel_prefix}-{tm_channel_suffix_from_client}"
+
+        assert to_client_chan is None or redis_server is not None, (
+            "You must specify a Redis server client object with a"
+            " RedisChannelSubscriber object"
+        )
+        self.to_client_chan = to_client_chan
+
+        if logger is None:
+            self.logger = logging.getLogger("tool-meister-client")
+        else:
+            self.logger = logger
+
+    def __enter__(self):
+        """On context entry, setup the connection with the Redis server.
+        """
+        if self.redis_server is None:
+            self.logger.debug("constructing Redis() object")
+            try:
+                self.redis_server = redis.Redis(
+                    host=self.redis_host, port=self.redis_port, db=0
+                )
+            except Exception as e:
+                self.logger.error(
+                    "Unable to connect to redis server, %s:%s: %s",
+                    self.redis_host,
+                    self.redis_port,
+                    e,
+                )
+                return 2
+            else:
+                self.logger.debug("constructed Redis() object")
+
+        if self.to_client_chan is None:
+            self.to_client_chan = RedisChannelSubscriber(
+                self.redis_server, self.to_client_channel
+            )
+        return self
+
+    def __exit__(self, *args):
+        """On context exit, just close down the to client channel object.
+        """
+        if self.to_client_chan is not None:
```

Comment:
Consider not closing the Redis channel if provided by the caller to allow them to do it.

Reply:
https://github.com/distributed-system-analysis/pbench/issues/2159